jrt/src/com/yahoo/jrt/MaybeTlsCryptoSocket.java

```diff
@@ -0,0 +1,126 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.jrt;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.SocketChannel;
+
+/**
+ * A crypto socket for the server side of a connection that
+ * auto-detects whether the connection is tls encrypted or unencrypted
+ * using clever heuristics. The assumption is that the client side
+ * will send at least one RPC request before expecting anything from
+ * the server. The first 9 bytes are inspected to see if they look
+ * like part of a tls handshake or not (RPC packet headers are 12
+ * bytes).
+ **/
+public class MaybeTlsCryptoSocket implements CryptoSocket {
+
+    private static final int snoop_size = 9;
+
+    private CryptoSocket socket;
+
+    // 'data' is the first 9 bytes received from the client
+    public static boolean looksLikeTlsToMe(byte[] data) {
+        if (data.length != snoop_size) {
+            return false; // wrong data size for tls detection
+        }
+        if (data[0] != 22) {
+            return false; // not tagged as tls handshake
+        }
+        if (data[1] != 3) {
+            return false; // unknown major version
+        }
+        if ((data[2] != 1) && (data[2] != 3)) {
+            return false; // unknown minor version
+        }
+        int frame_len = (data[3] & 0xff);
+        frame_len = ((frame_len << 8) | (data[4] & 0xff));
+        if (frame_len > (16384 + 2048)) {
+            return false; // frame too large
+        }
+        if (frame_len < 4) {
+            return false; // frame too small
+        }
+        if (data[5] != 0x1) {
+            return false; // not tagges as client hello
+        }
+        int hello_len = (data[6] & 0xff);
+        hello_len = ((hello_len << 8) | (data[7] & 0xff));
+        hello_len = ((hello_len << 8) | (data[8] & 0xff));
+        if ((frame_len - 4) != hello_len) {
+            return false; // inconsistent sizes; frame vs client hello
+        }
+        return true;
+    }
+
+    private class MyCryptoSocket extends NullCryptoSocket {
+
+        private TlsCryptoEngine factory;
+        private Buffer buffer;
+
+        MyCryptoSocket(SocketChannel channel, TlsCryptoEngine factory) {
+            super(channel);
+            this.factory = factory;
+            this.buffer = new Buffer(4096);
+        }
+
+        @Override public HandshakeResult handshake() throws IOException {
+            if (factory != null) {
+                channel().read(buffer.getWritable(snoop_size));
+                if (buffer.bytes() < snoop_size) {
+                    return HandshakeResult.NEED_READ;
+                }
+                byte[] data = new byte[snoop_size];
+                ByteBuffer src = buffer.getReadable();
+                for (int i = 0; i < snoop_size; i++) {
```

Comment:
Consider replacing loop with single call to `ByteBuffer.read(byte[] dst)`

Reply:
that would break the code by actually reading the data from the buffer (we need to only snoop it)